### PR TITLE
Add support for PSBT roles

### PR DESCRIPTION
**Pull request by t-bast**
We implement low-level support for all PSBT roles:

- [x] create
- [x] update
- [x] sign
- [x] combine
- [x] join
- [x] finalize
- [x] extract

We don't implement any smart wallet strategy to automatically find inputs we can sign for and sign them, because it's quite involved and can be dangerous (attackers could trick us into signing utxos we don't want to spend).

We instead let upper layers (eclair) build the scripts and witnesses based on their knowledge of what utxos belong to them.

### CODE REVIEW

File: src/main/scala/fr/acinq/bitcoin/Psbt.scala [hunk 1]
```diff
@@ -14,12 +14,186 @@ import scala.util.{Failure, Success, Try}
...
+          }
+          withUtxo.copy(
+            sighashType = sighashType.orElse(input.sighashType),
+            derivationPaths = input.derivationPaths ++ derivationPaths
+          )
+        case (_, input) => input // we don't update other inputs
+      }
+      Success(this.copy(inputs = updatedInputs))
+    }
+  }
```
Comment (sstone) [inline src/main/scala/fr/acinq/bitcoin/Psbt.scala:65]: I think that `update` would be more explicit if it was implemented with `Seq.indexWhere()` and `Seq.updated()`, and with a `Try` that around a simple method, for example:
```scala
def update(inputTx: Transaction,
             outputIndex: Int,
             redeemScript: Option[Seq[ScriptElt]] = None,
             witnessScript: Option[Seq[ScriptElt]] = None,
             sighashType: Option[Int] = None,
             derivationPaths: Map[PublicKey, KeyPathWithMaster] = Map.empty): Try[Psbt] = Try {

    require(outputIndex < inputTx.txOut.size, "output index must exist in the input tx")
    val outpoint = OutPoint(inputTx, outputIndex)
    val inputIndex = global.tx.txIn.indexWhere(_.outPoint == outpoint)
    require(inputIndex >= 0, "sbt transaction does not spend the provided outpoint")
    val input = this.inputs(inputIndex)
    val withUtxo = if (witnessScript.nonEmpty) {
      input.copy(witnessUtxo = Some(inputTx.txOut(outputIndex)), redeemScript = redeemScript.orElse(input.redeemScript), witnessScript = witnessScript)
    } else {
      input.copy(nonWitnessUtxo = Some(inputTx), redeemScript = redeemScript.orElse(input.redeemScript))
    }
    val withUtxo1 = withUtxo.copy(
      sighashType = sighashType.orElse(input.sighashType),
      derivationPaths = input.derivationPaths ++ derivationPaths
    )
    this.copy(inputs = this.inputs.updated(inputIndex, withUtxo1))
  }
```
WDYT ?

Same for `sign()`

Comment (t-bast) [reply src/main/scala/fr/acinq/bitcoin/Psbt.scala:59]: Good idea, done in https://github.com/ACINQ/bitcoin-lib/pull/43/commits/10066c9b5c63812bad463c0405583dbb07aa0bf7

File: src/main/scala/fr/acinq/bitcoin/Psbt.scala [hunk 2]
```diff
@@ -80,12 +242,16 @@ object Psbt {
...
-                                  witnessScript: Option[List[ScriptElt]],
-                                  scriptSig: Option[List[ScriptElt]],
+                                  redeemScript: Option[Seq[ScriptElt]],
+                                  witnessScript: Option[Seq[ScriptElt]],
+                                  scriptSig: Option[Seq[ScriptElt]],
                                   scriptWitness: Option[ScriptWitness],
                                   unknown: Seq[DataEntry]) extends DataMap
 
+  object PartiallySignedInput {
+    def empty: PartiallySignedInput = PartiallySignedInput(None, None, None, Map.empty, Map.empty, None, None, None, None, Nil)
```
Comment (sstone) [inline src/main/scala/fr/acinq/bitcoin/Psbt.scala:252]: Nit: why  a `def` and not a `val` ?

Comment (t-bast) [reply src/main/scala/fr/acinq/bitcoin/Psbt.scala:252]: No reason really, updating that.

Comment (t-bast) [reply src/main/scala/fr/acinq/bitcoin/Psbt.scala]: Done in https://github.com/ACINQ/bitcoin-lib/pull/43/commits/79ecd0f68e12a909a485f318fcac1700fb6a8db6

**Review by sstone — APPROVED** [review]

**Pull request merged**